File: 630 Anagrams (II)/solution.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StreamTokenizer st = new StreamTokenizer(br);
		st.nextToken();
		int T = (int)st.nval;
		StringBuilder output = new StringBuilder();
		while(T-- > 0) {
		    st.nextToken();
		    int N = (int)st.nval;
		    Anagram[] anagrams = new Anagram[N];
		    for(int i = 0; i < N; ++i) {
		        anagrams[i] = new Anagram();
		        st.nextToken();
		        anagrams[i].word = st.sval;
		        char[] temp = anagrams[i].word.toCharArray();
		        Arrays.sort(temp);
		        anagrams[i].sortedWord = new String(temp);
		    }
		    String test;
		    while(st.nextToken() == StreamTokenizer.TT_WORD && !(test = st.sval).equals("END")) {
		        output.append("Anagrams for: ").append(test).append('\n');
		        char[] temp = test.toCharArray();
		        Arrays.sort(temp);
		        String sortedTest = new String(temp);
		        int count = 0;
		        for(Anagram a : anagrams) {
		            if(sortedTest.equals(a.sortedWord))
		                output.append("  ").append(++count).append(") ").append(a.word).append('\n');
		        }
		        if(count == 0)
		            output.append("No anagrams for: ").append(test).append('\n');
		    }
		    if(T > 0)
		        output.append('\n');
		}
		System.out.print(output);
	}
}

class Anagram {
    String word, sortedWord;
}

File: 630 Anagrams (II)/solution.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

typedef struct {
    char word[21], sortedWord[21];
} Anagram;

int compare(const void *a, const void *b) {
    return *(char*)a - *(char*)b;
}

int main() {
	int T;
	scanf("%d", &T);
	bool firstCase = true;
	while(T--) {
	    if(!firstCase)
	        putchar('\n');
	    int N;
	    scanf("%d", &N);
	    Anagram anagrams[N];
	    for(int i = 0; i < N; ++i) {
	        scanf("%s", anagrams[i].word);
	        strcpy(anagrams[i].sortedWord, anagrams[i].word);
	        int len = strlen(anagrams[i].sortedWord);
	        qsort(anagrams[i].sortedWord, len, sizeof(char), compare);
	    }
	    char test[21], sortedTest[21];
	    while(scanf("%s", test) && strcmp(test, "END") != 0) {
	        printf("Anagrams for: %s\n", test);
	        strcpy(sortedTest, test);
	        int len = strlen(sortedTest);
	        qsort(sortedTest, len, sizeof(char), compare);
	        int count = 0;
	        for(int i = 0; i < N; ++i) {
	            if(strcmp(sortedTest, anagrams[i].sortedWord) == 0)
	                printf("  %d) %s\n", ++count, anagrams[i].word);
	        }
	        if(count == 0)
	            printf("No anagrams for: %s\n", test);
	    }
	    firstCase = false;
	}
	return 0;
}
